crates/warg-client/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+use semver::Version;
+use url::Url;
+
+/// A client for interacting with a registry
+#[allow(dead_code)]
+pub struct Client<RegistryCache, ValidationCache> {
+    registry: DomainName,
+    known_roots: Vec<Root>,
+    registry_cache: RegistryCache,
+    validation_cache: ValidationCache,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DomainName(String);
+
+#[allow(dead_code)]
+pub struct Root {
+    bytes: Vec<u8>
+}
+
+pub trait Cache {
+    /// Insert a value into the clients cache.
+    fn insert(&mut self, key: Vec<u8>, value: Vec<u8>);
+
+    /// Lookup a previously inserted cache value.
+    fn lookup(&self, key: &[u8]) -> Option<Vec<u8>>;
```

Comment:
Probably better to return `Option<&[u8]>` and let the consumer clone if necessary.